libkbfs/bserver_tlf_journal.go

```diff
@@ -0,0 +1,706 @@
+	[]byte, BlockCryptKeyServerHalf, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.getDataLocked(id, context)
+}
+
+func (s *bserverTlfJournal) getAll() (
+	map[BlockID]map[BlockRefNonce]blockRefLocalStatus, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if s.isShutdown {
+		return nil, errBserverTlfJournalShutdown
+	}
+
+	res := make(map[BlockID]map[BlockRefNonce]blockRefLocalStatus)
+
+	for id, refs := range s.refs {
+		if len(refs) == 0 {
+			continue
+		}
+
+		res[id] = make(map[BlockRefNonce]blockRefLocalStatus)
+		for ref, refEntry := range refs {
+			res[id][ref] = refEntry.Status
+		}
+	}
+
+	return res, nil
+}
+
+func (s *bserverTlfJournal) putData(
+	id BlockID, context BlockContext, buf []byte,
+	serverHalf BlockCryptKeyServerHalf) error {
+	err := validateBlockServerPut(s.crypto, id, context, buf)
+	if err != nil {
+		return err
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isShutdown {
+		return errBserverTlfJournalShutdown
+	}
+
+	_, existingServerHalf, err := s.getDataLocked(id, context)
+	var exists bool
+	switch err.(type) {
+	case BServerErrorBlockNonExistent:
+		exists = false
+	case nil:
+		exists = true
+	default:
+		return err
+	}
+
+	if exists {
+		// If the entry already exists, everything should be
+		// the same, except for possibly additional
+		// references.
+
+		// We checked that both buf and existingData hash to
+		// id, so no need to check that they're both equal.
+
+		if existingServerHalf != serverHalf {
+			return fmt.Errorf(
+				"key server half mismatch: expected %s, got %s",
+				existingServerHalf, serverHalf)
+		}
+	}
+
+	err = os.MkdirAll(s.blockPath(id), 0700)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(s.blockDataPath(id), buf, 0600)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Add integrity-checking for key server half?
+
+	err = ioutil.WriteFile(
+		s.keyServerHalfPath(id), serverHalf.data[:], 0600)
+	if err != nil {
+		return err
+	}
+
+	err = s.putRefEntryLocked(id, blockRefEntry{
+		Status:  liveBlockRef,
+		Context: context,
+	})
+	if err != nil {
+		return err
+	}
+
+	return s.appendJournalEntryLocked("put", id, []BlockContext{context})
+}
+
+func (s *bserverTlfJournal) addReference(id BlockID, context BlockContext) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isShutdown {
+		return errBserverTlfJournalShutdown
+	}
+
+	refs := s.refs[id]
+	if refs == nil {
+		return BServerErrorBlockNonExistent{fmt.Sprintf("Block ID %s "+
+			"doesn't exist and cannot be referenced.", id)}
+	}
+
+	// Only add it if there's a non-archived reference.
+	hasNonArchivedRef := false
+	for _, refEntry := range refs {
+		if refEntry.Status == liveBlockRef {
+			hasNonArchivedRef = true
+			break
+		}
+	}
+
+	if !hasNonArchivedRef {
+		return BServerErrorBlockArchived{fmt.Sprintf("Block ID %s has "+
+			"been archived and cannot be referenced.", id)}
+	}
+
+	// TODO: Allow adding a reference even if all the existing
```

Comment:
Yeah, I came to the same conclusion. I think it suffices to make `BlockServer.AddReference` take the `data`/`keyServerHalf` arguments, so that the journal can stash them. I'll file a story for that.

I'm surprised to hear that re-putting the same block with the same ID will be hard. Why is that? I'll file a story for that, too.